Extend `/incidents` endpoint to handle `POST` requests as well (#456)
Icinga/icinga-notifications: doc/20-HTTP-API.md

    curl must be executed as a user which is configured as listener_username of a source.

## Incidents

The `/incidents` Icinga Notifications HTTP API endpoint allows sources to query and modify a list of open incidents

> belonging to its objects. You can retrieve incidents by sending a `GET` request to the endpoint, and you can modify

Reviewer: > You can

Please keep the description in the passive voice as the rest of the documentation. Same for the following paragraphs. For example, consult the _Process Event_ subsection above.

Author: As discussed offline, I still don't know why we can't use active voice for our docs as for example the [`Postgres` docs](https://www.postgresql.org/docs/current/sql-select.html) does but changed it anyway.

Authentication follows the same transport-specific rules as for event submission: TCP requires HTTP
Basic Auth with username and password, while a Unix socket identifies the caller by their OS user.

### Getting Incidents

In order to retrieve incidents, you can send a `GET` request to the `/incidents` endpoint with the appropriate `filter`
query parameter. Currently, this endpoint will include the attributes listed below in the response for each incident:

| Attribute   | Description                                                                |
|-------------|----------------------------------------------------------------------------|
| is_muted    | A boolean indicating whether the incident is muted or not.                 |
| object_tags | A dictionary containing the object ID tags associated with the incident.   |
| severity    | The severity level of the incident (e.g., `crit`, `err`, `warning`, etc.). |

For instance, when using Icinga DB as a source, the `environment` object ID tag can be used to filter incidents for a
specific Icinga DB environment. The following example shows how to retrieve all incidents for the
`08434a503ec43bb67cd380c5d0b6217a1ebf924b` environment:

> 

Reviewer: Consider showing a human-readable JSON filter first, as the URL-encoded filter is quite hard to read. Same for the following, even more cryptic filter queries.

Author: I've already referenced the `API Filtering` section as part of the `Incidents`, so repeating everything here doesn't make any sense to me. 

> Same for the following, even more cryptic filter queries.

If you are referring to the modify incidents example, then I already have provided the human-readable version of it as you already [commented it.](https://github.com/Icinga/icinga-notifications/pull/456#discussion_r3528092002)

]
```

### Modifying Incidents

Modifying incidents can be done by sending a `POST` request to the `/incidents` endpoint with the appropriate
`filter` query parameter and a JSON-encoded body that describes the desired changes. The request body should be a JSON

> object that contains the attributes to be modified set to their new values. The list of attributes that currently can

Reviewer: ```suggestion
object that contains the attributes to be modified, set to their new values. The list of attributes that currently can
```

Author: Fixed.

be modified are listed in the following table.

| Attribute | Description                                                                        |
|-----------|------------------------------------------------------------------------------------|
| close     | Closes the incident matching the filter. This attribute can only be set to `true`. |
| message   | Updates the incident's message. This attribute can be set to any string value.     |

As opposed to the `/process-event` endpoint, closing an incident this way won't have any side effects other than marking

> the incident as closed. For instance, it won't trigger any notifications or trigger any escalations. Also, note that

Reviewer: ```suggestion
As opposed to the `/process-event` endpoint, closing an incident this way will not have any side effects other than marking
the incident as closed. For instance, it will not trigger any notifications or trigger any escalations. Also, note that
```

No contractions in formal writings.

Author: > No contractions in formal writings.

Hu, since when?

```
~/Workspace/go/icingadb (main ✗) grep -Rn "doesn't" doc
doc/03-Configuration.md:234:`/etc/icingadb/config.yml` file to disable it entirely, as Icinga DB doesn't have to wait for cluster
doc/02-Installation.md.d/For-Container.md:39:The container doesn't need any volumes and takes the environment variables.
~/Workspace/go/icingadb (main ✗) grep -Rn "won't" doc
doc/06-Migration.md:87:process won't be interrupted by a reboot (of the machine
~/Workspace/go/icingadb (main ✗) grep -Rn "isn't" doc
doc/04-Upgrading.md:112:the schema upgrade required to fix the history tables isn't automatically applied by `1.2.0.sql` as a rewrite of the
doc/06-Migration.md:86:internal cache. If either there isn't much to migrate or the migration

---

~/Workspace/go/icinga-notifications (extend-incidents-endpoint ✗) grep -Rn "doesn't" doc
doc/20-HTTP-API.md:265:might have noticed, the filter syntax doesn't support more complex logical expressions, such as nested conditions or
doc/03-Configuration.md:246:`/etc/icinga-notifications/config.yml` file to disable it entirely, as Icinga Notifications doesn't have to wait for cluster
```

https://docs.netways.de/books/development/page/pull-request-und-code-review-guidelines doesn't metnion it either. I'm not too attached to this in anyway but this isn't something that we all have respected thus far. Nevertheless, I'm going to change this...

Reviewer: > Hu, since when?

Dunno, I always tried it to keep it that way as I learned it for formal writing. See the following classification of the Cambridge Dictionary: https://dictionary.cambridge.org/grammar/british-grammar/contractions

Author: Changed it.

the `close` attribute can only be set to `true`, otherwise Icinga Notifications will reject the request with a 400
status code. The `message` attribute can be used to update the incident's message without any side effects, as well.
For instance, this can be useful to regularly synchronize only the plugin output of the associated object so that the
incident's message is always up to date.

The following example shows how to close the incident for the `mailserver` host and `filesystem` service in the
`08434a503ec43bb67cd380c5d0b6217a1ebf924b` environment:

```

> $ curl -u 'example:insecureinsecure' -X POST 'http://localhost:5680/incidents?filter=%7B%22environment%22%3A%2208434a503ec43bb67cd380c5d0b6217a1ebf924b%22%2C%22host%22%3A%22mailserver%22%2C%22service%22%3A%22filesystem%22%7D' d '@-' <<EOF

Reviewer: ```suggestion
$ curl -u 'example:insecureinsecure' -X POST 'http://localhost:5680/incidents?filter=%7B%22environment%22%3A%2208434a503ec43bb67cd380c5d0b6217a1ebf924b%22%2C%22host%22%3A%22mailserver%22%2C%22service%22%3A%22filesystem%22%7D' -d '@-' <<EOF
```

Author: Fixed.

{
  "close": true
}
EOF
```

The URL-encoded filter in the example above corresponds to the following JSON object:

```json
{
  "environment": "08434a503ec43bb67cd380c5d0b6217a1ebf924b",
  "host": "mailserver",
  "service": "filesystem"
}

> ```

Reviewer: Consider showing the human-readable version before the `curl(1)` example, as otherwise the reader first sees the binary soup and is either confused or decodes it manually.

Author: Moved it.

If you want to update the message of an incident instead of closing it, you can send a `POST` request with the `message`
attribute set to the new message. For instance, the following example shows how to update the message of the incident
for the `database` host and `load` service in the `08434a503ec43bb67cd380c5d0b6217a1ebf924b` environment:

```

> $ curl -u 'example:insecureinsecure' -X POST 'http://localhost:5680/incidents?filter=%7B%22environment%22%3A%2208434a503ec43bb67cd380c5d0b6217a1ebf924b%22%2C%22host%22%3A%22database%22%2C%22service%22%3A%22load%22%7D' d '@-' <<EOF

Reviewer: ```suggestion
$ curl -u 'example:insecureinsecure' -X POST 'http://localhost:5680/incidents?filter=%7B%22environment%22%3A%2208434a503ec43bb67cd380c5d0b6217a1ebf924b%22%2C%22host%22%3A%22database%22%2C%22service%22%3A%22load%22%7D' -d '@-' <<EOF
```

Author: Fixed.

{
  "message": "The load on the database server has returned to normal."
}
EOF
```

## API Filtering

Some of the Icinga Notifications API endpoints require to explicitly specify which objects the request should be
filtered for. For instance, the `/incidents` endpoint requires a `filter` query parameter to specify which incidents
you want to either retrieve or modify. The `filter` query parameter is a JSON-encoded object that describes the
filtering criteria. The filter syntax and semantics are as follows:

> The `filter` parameter can either be a JSON object or a JSON array of objects. Each object represents a set of filtering

Reviewer: ```suggestion
The `filter` parameter can either be a JSON object or a non-empty JSON array of objects. Each object represents a set of filtering
```

Otherwise, `[]` would be a valid filter, matching everything. If I understood the gist of #452 correctly, this should not be allowed, right?

Furthermore, the current code already rejects this (empty slice, `matched` stays `false`).

Author: > Otherwise, `[]` would be a valid filter, matching everything.

It's a valid filter, but how do you came to the _matching everything_ part?

> Furthermore, the current code already rejects this (empty slice, `matched` stays `false`).

It doesn't reject it but the filter simply evaluates to false because nothing was provided that would match on any of the incidents, so I don't know why you actually think that this is a _match everything_ thing. It already states _a JSON array of objects_, which is non-empty by definition but I don't mind adding the non-empty part too, though that doesn't change the fact that when you provide `[]` as a filter, it will still evaluate to false.

Reviewer: > It's a valid filter, but how do you came to the matching everything part?

Following the vacuous truth of the empty set in set theory.

> It already states a JSON array of objects, which is non-empty by definition

I would kindly disagree. An empty array would satisfy this requirement, being an array with zero objects, the empty set. Following this reasoning and the vacuous truth attribute from above, `[]` would be a match-all.

---

But since this is documented now, it's fine for me :)

criteria that must be satisfied for an incident to be included in the response. All the key value pairs in the object

> are combined with a logical AND, meaning that all the specified criteria must be met. For instance, given the following

Reviewer: Please state that the evaluation of objects within an array is combined using a logical OR. After reading the first chapter, I only had an AND match in my head, so I was quite confused when I got to the first array example, even though the OR is mentioned there.

Author: This section exclusively explains how the JSON object is evaluated, how and why should I mention the evaluation of JSON arrays here? I don't actually get how this even confuses you since it explicitly states there that the JSON object key value pairs are evaluated with a logical AND. This looks fine to me, but you can suggest an alternative sentence that would fit here without looking completely out of place.

```json
{
  "host": "mailserver",
  "service": null
}
```

This filter will match incidents that have the `host` object ID tag set to `mailserver` and do **not** have a `service`
object ID tag at all. In other words, the `service` tag must not exist in the incident's object tags for the filter to
be satisfied. Now, if you want to express a logical OR condition, you can use a JSON array of objects. For example,
given the following filter:

```json
[
  {
    "host": "mailserver",
    "service": "filesystem"
  },
  {
    "host": "database",
    "service": null
  }
]
```

This filter will match incidents that either have the `host` tag set to `mailserver` **and** the `service` tag set to
`filesystem`, **or** have the `host` tag set to `database` and do **not** have a `service` tag at all. In other words,
the filter will match incidents that satisfy either of the two sets of criteria, allowing for more complex filtering
scenarios. These filtering capabilities enable you to precisely target the incidents you want to retrieve or modify
based on their object tags, providing flexibility and control over the incident management process. However, as you

> might have noticed, the filter syntax doesn't support more complex logical expressions, such as nested conditions or

Reviewer: ```suggestion
might have noticed, the filter syntax does not support more complex logical expressions, such as nested conditions or
```

Author: Changed.

combinations of `AND` and `OR` within a single object. Any attempt to use such expressions will result in a
`400 Bad Request` response, indicating that the filter is invalid.

> When constructing filters, it's important to ensure that the filter is properly URL-encoded when included in the query

Reviewer: ```suggestion
When constructing filters, it is important to ensure that the filter is properly URL-encoded when included in the query
```

Author: Changed.

```
$ curl -u 'example:insecureinsecure' -X POST 'http://localhost:5680/incidents?filter=%7B%22environment%22%3A%2208434a503ec43bb67cd380c5d0b6217a1ebf924b%22%2C%22host%22%3A%22database%22%2C%22service%22%3A%22load%22%7D' -d '@-' <<EOF
{
  "message": "The load on the database server has returned to normal."
}
EOF
```

When bulk modifying incidents, the changes will be applied to all the matching incidents that satisfy the filter
sequentially. If any of the incidents cannot be modified due to some reason, each incident will convey its own status
in the response, and the general HTTP status code will be the most severe one among all the incidents. For instance, if
one incident is successfully closed, but another one cannot, the response will contain a `200 OK` status code for the
first incident and a `500` status code for the second incident, and the overall HTTP status code will be
`500 Internal Server Error`. The complete response body looks like this:

```json
[
  {
    "object_tags": {
      "environment": "08434a503ec43bb67cd380c5d0b6217a1ebf924b",
      "host": "mailserver",
      "service": "filesystem"
    },
    "code": 200,

>     "status": "Incident modified successfully."

Reviewer: ```suggestion
    "status": "incident modified successfully"
```

https://github.com/Icinga/icinga-notifications/blob/7f15d056aab6c59c2c3aca3f8970447a3378357f/internal/listener/listener.go#L506-L508

Users might end up asserting the exact status message from the docs. Not that they should, but maybe someone is going to.

Author: Fixed.
